File: typescript/src/modules/notifier/NotifierModule.ts
import type { IModule } from '../../core/IModule';
import type { IEventBus } from '../../core/EventBus';
import type { IModuleLifecycle } from '../../core/Lifecycle';

// ── Events ──
export interface CalculationPerformedEvent {
    operation: string;
    result: string;
}

export interface GreetingEvent {
    name: string;
    message: string;
}

/**
 * NotifierModule — Demo EventBus + Lifecycle.
 * Subscribes to events without importing source modules.
 */
export class NotifierModule implements IModule, IModuleLifecycle {
    private readonly notifications: string[] = [];

    constructor(private readonly eventBus: IEventBus) { }

    readonly name = 'notifier';
    readonly description = '🔔 Listens to events from other modules';
    readonly commands = ['history', 'count', 'clear'] as const;

    execute(command: string, _args: string[]): string {
        switch (command.toLowerCase()) {
            case 'history':
                if (this.notifications.length === 0) return '📭 No notifications yet.';
                return this.notifications.slice(-10)
                    .map((n, i) => `  ${i + 1}. ${n}`).join('\n');
            case 'count':
                return `📬 ${this.notifications.length} notification(s)`;
            case 'clear': {
                const count = this.notifications.length;
                this.notifications.length = 0;
                return `🗑️ Cleared ${count} notification(s)`;
            }
            default:
                return `❓ Unknown command: ${command}`;
        }
    }

    // Lifecycle
    async onInitializing(): Promise<void> { }
    async onInitialized(): Promise<void> {
        this.eventBus.subscribe<CalculationPerformedEvent>('CalculationPerformed', (e) => {
            this.notifications.push(`🧮 Calculation: ${e.operation} = ${e.result}`);
        });
        this.eventBus.subscribe<GreetingEvent>('Greeting', (e) => {
            this.notifications.push(`👋 Greeting: ${e.message}`);
        });
    }
    async onShuttingDown(): Promise<void> { }
    async onShutdown(): Promise<void> { }
}

File: csharp/Modules/Notifier/NotifierModule.cs
using UniverseDemo.Core;

namespace UniverseDemo.Modules.Notifier;

// ── Events ──

/// <summary>Event khi một phép tính được thực hiện.</summary>
public sealed class CalculationPerformedEvent
{
    public required string Operation { get; init; }
    public required string Result { get; init; }
    public DateTimeOffset Timestamp { get; init; } = DateTimeOffset.UtcNow;
}

/// <summary>Event khi có lời chào.</summary>
public sealed class GreetingEvent
{
    public required string Name { get; init; }
    public required string Message { get; init; }
}

/// <summary>
/// NotifierModule — Module lắng nghe events từ modules khác.
/// Demo Principle #5: Indirect Communication via EventBus.
/// Không import Calculator hay Greeter — chỉ subscribe events.
/// </summary>
public sealed class NotifierModule : IModule, IModuleLifecycle
{
    private readonly IEventBus _eventBus;
    private readonly List<string> _notifications = [];
    private bool _initialized;

    public NotifierModule(IEventBus eventBus)
    {
        _eventBus = eventBus;
    }

    public string Name => "notifier";
    public string Description => "🔔 Listens to events from other modules";
    public IReadOnlyList<string> Commands => ["history", "count", "clear"];

    public string Execute(string command, string[] args) => command.ToLower() switch
    {
        "history" => _notifications.Count == 0
            ? "📭 No notifications yet."
            : string.Join("\n", _notifications.TakeLast(10).Select((n, i) => $"  {i + 1}. {n}")),
        "count" => $"📬 {_notifications.Count} notification(s)",
        "clear" => ClearNotifications(),
        _ => $"❓ Unknown command: {command}"
    };

    // ── Lifecycle ──

    public Task OnInitializing() => Task.CompletedTask;

    public Task OnInitialized()
    {
        // Subscribe events SAU khi đã đăng ký
        _eventBus.Subscribe<CalculationPerformedEvent>(OnCalculation);
        _eventBus.Subscribe<GreetingEvent>(OnGreeting);
        _initialized = true;
        return Task.CompletedTask;
    }

    public Task OnShuttingDown()
    {
        _eventBus.Unsubscribe<CalculationPerformedEvent>(OnCalculation);
        _eventBus.Unsubscribe<GreetingEvent>(OnGreeting);
        return Task.CompletedTask;
    }

    public Task OnShutdown()
    {
        _initialized = false;
        return Task.CompletedTask;
    }

    // ── Event Handlers ──

    private void OnCalculation(CalculationPerformedEvent e)
    {
        _notifications.Add($"🧮 Calculation: {e.Operation} = {e.Result}");
    }

    private void OnGreeting(GreetingEvent e)
    {
        _notifications.Add($"👋 Greeting: {e.Message}");
    }

    private string ClearNotifications()
    {
        var count = _notifications.Count;
        _notifications.Clear();
        return $"🗑️ Cleared {count} notification(s)";
    }
}
